### PR TITLE
[READY] Drones die when trying to leave the derelict

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

Drones self destruct when leaving the derelict (changing zlevels)

## Why It's Good For The Game

This was requested by an admin. You can slap everything in someone's face to NOT LEAVE THE DERELICT, and they still will. So here.

## Changelog
:cl:
admin: drones are no longer allowed to leave their station, they will die upon doing so
/:cl:

### CODE REVIEW

**Comment by TheVekter** [comment]
When you tell people not to do something, they will try and find every single way to do it regardless.

**Comment by ATH1909** [comment]
Time to instakill drones by teleporting them off of the z-level.

Could you make it teleport them instead of kill them? You already have a parameter for that; just set it to false (I feel like that'd be a lot less harsh, especially if a drone accidentally starts drifting off into space or something).

**Comment by ATH1909** [comment]
Time to instakill drones by teleporting them off of the z-level.

Could you make it teleport them instead of kill them? You already have a parameter for that; just set it to false (I feel like that'd be a lot less harsh, especially if a drone accidentally drifts off into space or something).

**Comment by kriskog** [comment]
>
>
> Time to instakill drones by teleporting them off of the z-level.

Admin issue or rule 4

File: tgstation.dme [hunk 1]
```diff
@@ -414,7 +414,9 @@
 #include "code\datums\components\spill.dm"
 #include "code\datums\components\spooky.dm"
 #include "code\datums\components\squeak.dm"
+#include "code\datums\components\stationstuck.dm"
```
Comment (spookydonut) [inline tgstation.dme]:
```suggestion
```

**Comment by kriskog** [comment]
Reminder that this is the greeting derelict drones get. People that leave willingly deserve deactivation.
image

File: code/datums/components/stationstuck.dm [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	if(original)
+		if(istype(newc))
+			murder = newc.murder
+			teleport_message = newc.teleport_message
+			death_message = newc.death_message
+
+/datum/component/stationstuck/proc/punish()
+	var/mob/living/L = parent
+	if(murder)
+		to_chat("<span class='userdanger'>[death_message]</span>")
```
Comment (tralezab) [inline code/datums/components/stationstuck.dm]: so that's why those darn messages aren't working

**Comment by zxaber** [comment]
What happens if someone brings the unactivated drone to the main station? Does it explode as soon as a ghost takes it?

**Comment by kriskog** [comment]
The drone shells are anchored, so should not be possible.

**Comment by MCterra10** [comment]
This is great, except the one issue I can see is accidentally floating away and dying because drones can't throw stuff to change direction

**Comment by kriskog** [comment]
When I made upgrades to the derelict, I also added lattices to most of the corners to prevent most accidental floating off situations. Some are still missing on the solars at the very edges of the map though, and lag could play into it so its not impossible to accidentally float off.

Give the derelict drones a pocket fire extinguisher in their bags as they spawn, that should get them home in a pinch. Otherwise, if the drone shell printer is on they can take a new shell and carry on.

**Comment by wesoda25** [comment]
who tf is downvoting this??????

**Comment by plapatin** [comment]
> 
> 
> This is great, except the one issue I can see is accidentally floating away and dying because drones can't throw stuff to change direction

a smart drone would carry rods in their bag (which they probably already have anyways, because they're building lots of stuff) to build lattices in case they float off

**Comment by tralezab** [comment]
i can throw in a fire extinguisher

File: code/modules/mob/living/simple_animal/friendly/drone/extra_drone_types.dm [hunk 3]
```diff
@@ -124,3 +124,11 @@
 	"<span class='notice'>     - Interacting with non-drone players outside KS13, dead or alive.</span>\n"+\
 	"<span class='warning'>These rules are at admin discretion and will be heavily enforced.</span>\n"+\
 	"<span class='warning'><u>If you do not have the regular drone laws, follow your laws to the best of your ability.</u></span>"
+
+/mob/living/simple_animal/drone/derelict/Initialize()
+	..()
```
Comment (ninjanomnom) [inline code/modules/mob/living/simple_animal/friendly/drone/extra_drone_types.dm]:
```suggestion
	. = ..()
```

File: code/datums/components/stationstuck.dm [hunk 4]
```diff
@@ -0,0 +1,41 @@
+//very similar to stationloving, but more made for mobs and not objects. used on derelict drones currently
+/datum/component/stationstuck
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+	var/murder = TRUE //teleports if not
+	var/stuck_zlevel
+	var/teleport_message = ""
+	var/death_message = ""
+
+/datum/component/stationstuck/Initialize(_murder = TRUE, _teleport_message = "", _death_message = "")
```
Comment (ninjanomnom) [inline code/datums/components/stationstuck.dm]: You just need a single message argument since only one will ever be used depending on `_murder`

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/datums/components/stationstuck.dm (hunk 4 above)
Comment (tralezab) [reply code/datums/components/stationstuck.dm]: i like this idea

File: code/datums/components/stationstuck.dm [hunk 5]
```diff
@@ -0,0 +1,41 @@
+//very similar to stationloving, but more made for mobs and not objects. used on derelict drones currently
+/datum/component/stationstuck
```
Comment (optimumtact) [inline code/datums/components/stationstuck.dm]: can this not be a subtype of stationloving?

Comment (tralezab) [reply code/datums/components/stationstuck.dm]: stationloving having a clear purpose for station important objectives is good, having it also cram in mob code that then wouldn't act the same as stationloving is weird and bad

**Pull request merged**